### PR TITLE
SALTO-1145: fixed cli configuration options written to disk on suggestions

### PR DESCRIPTION
**Pull request by alonstern**
Fixed a bug where a CLI configuration option would be written to the configuration file when an adapter made a configuration suggestion (A more detailed explanation can be found in https://salto-io.atlassian.net/browse/SALTO-1145).

---

The solution of this PR is first, to check whether there is a conflict between the configuration changes made by the adapter and the changes made by the user. Second to apply the configuration changes made by the adapter on the configuration from the disk (without the overrides) to not include the changes made by the user.

---
_Release Notes_: 
Fixed a bug where a CLI configuration override (i.e., `-C` or `export SALTO_*`) would be written to the configuration file when an adapter made a configuration suggestion.

### CODE REVIEW

File: packages/core/src/local-workspace/workspace_config.ts [hunk 1]
```diff
@@ -82,8 +109,34 @@ export const workspaceConfigSource = async (
...
-      repoCs.set(path.join(ADAPTERS_CONFIG_NAME, adapter), config)
-    ,
+    setAdapter: async (adapter: string, config: Readonly<InstanceElement>): Promise<void> => {
+      const currConf = await repoCs.get(path.join(ADAPTERS_CONFIG_NAME, adapter))
+      // Could happen at the initialization of a service.
+      if (currConf === undefined) {
+        await repoCs.set(path.join(ADAPTERS_CONFIG_NAME, adapter), config)
+        return
+      }
+      const configChanges = await getPlan({ before: [currConf], after: [config.clone()] })
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/workspace_config.ts:119]: should use `detailedCompare` instead of a full-blown `getPlan`

File: packages/core/src/local-workspace/workspace_config.ts [hunk 2]
```diff
@@ -82,8 +109,34 @@ export const workspaceConfigSource = async (
...
+      )
+      if (currConfWithoutOverrides === undefined) {
+        // Shouldn't get here since we were able the receive the configuration with the overrides.
+        throw new Error('failed to set configuration. Could not access to get the current configuration')
+      }
+
+      const detailedChanges = wu(configChanges.itemsByEvalOrder())
+        .map(item => item.detailedChanges())
+        .flatten()
+      for (const change of detailedChanges) {
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/workspace_config.ts:135]: 1. style - we do not use for loops, we use `forEach` when there is no choice but to use side-effects
2. there is already a function called `applyDetailedChanges` for this

File: packages/core/src/local-workspace/workspace_config.ts [hunk 3]
```diff
@@ -82,8 +109,34 @@ export const workspaceConfigSource = async (
     },
     getAdapter: (adapter: string): Promise<InstanceElement | undefined> =>
       repoCs.get(path.join(ADAPTERS_CONFIG_NAME, adapter)),
-    setAdapter: async (adapter: string, config: Readonly<InstanceElement>): Promise<void> =>
-      repoCs.set(path.join(ADAPTERS_CONFIG_NAME, adapter), config)
-    ,
+    setAdapter: async (adapter: string, config: Readonly<InstanceElement>): Promise<void> => {
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/workspace_config.ts:112]: why do we need to do this at this level?
this is not the layer that deals with config overrides, which makes it work harder to implement something that could probably be done more easily if it was implemented at the layer that implements the config overrides in the first place

File: packages/core/src/local-workspace/workspace_config.ts [hunk 4]
```diff
@@ -48,6 +51,30 @@ export const workspaceConfigSource = async (
...
+      .map(item => item.detailedChanges())
+      .flatten()
+      .map(detailedChange => detailedChange.id.getFullName())
+      .toArray()
+
+    const overiddenConfigFields = configOverrides.map(override => override.id.getFullName())
+
+    const updatedOverriddenFields = overiddenConfigFields.filter(
+      overiddenField => updatedConfigFields.some(updatedField =>
+        updatedField.startsWith(overiddenField) || overiddenField.startsWith(updatedField))
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/workspace_config.ts:66]: `startsWith` is not a safe way to compare element IDs (would be true for `bla.Foo` and `bla.FooBar` even though that are not related)
that is why element ids have a `isParentOf` function.

File: packages/core/src/local-workspace/workspace_config.ts (hunk 2 above)
Comment (alonstern) [reply packages/core/src/local-workspace/workspace_config.ts:135]: Fixed, but why we don't use loops? 🤔

Comment (ori-moisis) [reply packages/core/src/local-workspace/workspace_config.ts:135]: style
we prefer functional code and generally most of our code is loosely inclined to "[fluent](https://nikas.praninskas.com/javascript/2015/04/26/fluent-javascript/)" principles

File: packages/workspace/src/workspace/config_source.ts [hunk 5]
```diff
@@ -39,49 +39,94 @@ class ConfigParseError extends Error {
...
+        name,
+        parseResult.elements.map(elem => elem.elemID.getFullName()).join(','),
+      )
+      return undefined
+    }
+    return configInstance
+  }
+
+  const applyConfigOverrides = (conf: InstanceElement): void => {
+    const overridesForInstance = configOverridesById[conf.elemID.adapter] ?? []
```
Comment (alonstern) [inline packages/workspace/src/workspace/config_source.ts:78]: @ori-moisis  shoud I also `applyDetailedChanges` here?

Comment (ori-moisis) [reply packages/workspace/src/workspace/config_source.ts:78]: I guess so, yes

File: packages/workspace/src/workspace/config_source.ts [hunk 6]
```diff
@@ -39,49 +39,94 @@ class ConfigParseError extends Error {
...
+    const updatedOverriddenIds = configOverrides.filter(
+      overiddeChange => configChanges.some(
+        updateChange => updateChange.id.isParentOf(overiddeChange.id)
+          || overiddeChange.id.isParentOf(updateChange.id)
+          || overiddeChange.id.isEqual(updateChange.id)
+      )
+    )
+
+    if (updatedOverriddenIds.length !== 0) {
+      throw new Error(`cannot update fields that were overriden by the user: ${updatedOverriddenIds.map(change => change.id)}`)
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/config_source.ts:99]:
```suggestion
      throw new Error(`cannot update fields that were overriden by the user: ${updatedOverriddenIds.map(change => change.id.getFullName())}`)
```
I assume the default string representation of the element id won't be very readable

File: packages/workspace/src/workspace/config_source.ts [hunk 7]
```diff
@@ -39,49 +39,94 @@ class ConfigParseError extends Error {
...
+      // Could happen at the initialization of a service.
+      if (currConfWithoutOverrides === undefined) {
+        await setUnsafe(name, config)
+        return
+      }
+
+      const currConf = currConfWithoutOverrides.clone()
+      applyConfigOverrides(currConf)
+
+      const configChanges = await detailedCompare(currConf, config.clone())
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/config_source.ts:123]: why the extra `clone`?

**Review by ori-moisis — APPROVED** [review]
